### PR TITLE
[Distributed] Fix a bug in parallel graph partition

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
ParMETIS cannot handle self-loops and duplicated edges correctly. Previous steps cannot generate graph partitions that contain all edges in the original input graph. This PR fixes this by removing the self-loops and duplicated edges and save them separately. After ParMETIS, we merge the removed edges back.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by VoVAllen — COMMENTED** [review]
I saw lots of numpy operations included, which might use lots of memory. How large can this PR scale to? Is there any part you think might be better implemented in DGL?

File: examples/pytorch/rgcn/experimental/verify_mag_partitions.py [hunk 1]
```diff
@@ -101,9 +102,15 @@
...
 
         # Check edge data.
         for name in hg.edges[etype].data:
             local_data = edge_feats[etype + '/' + name][type_eid]
             local_data1 = hg.edges[etype].data[name][orig_type_eid]
             assert np.all(local_data.numpy() == local_data1.numpy())
+
+for etype in orig_edge_ids:
+    eids = th.cat(orig_edge_ids[etype])
+    uniq_eids = th.unique(eids)
```
Comment (classicsong) [inline examples/pytorch/rgcn/experimental/verify_mag_partitions.py:115]: Why do we call unique here?
Are there any duplicated edge ids in eids?

File: examples/pytorch/rgcn/experimental/write_mag.py [hunk 2]
```diff
@@ -46,11 +42,43 @@
...
+self_loop_src_id = src_id[self_loop_idx]
+self_loop_dst_id = dst_id[self_loop_idx]
+self_loop_orig_id = g.edata['orig_id'][self_loop_idx]
+self_loop_etype = g.edata[dgl.ETYPE][self_loop_idx]
+src_id = src_id[not_self_loop_idx]
+dst_id = dst_id[not_self_loop_idx]
+orig_id = g.edata['orig_id'][not_self_loop_idx]
+etype = g.edata[dgl.ETYPE][not_self_loop_idx]
+# Remove duplicated edges.
+ids = (src_id * g.number_of_nodes() + dst_id).numpy()
```
Comment (classicsong) [inline examples/pytorch/rgcn/experimental/write_mag.py:60]: It is better to add some comments here.

File: examples/pytorch/rgcn/experimental/write_mag.py [hunk 3]
```diff
@@ -46,11 +42,43 @@
...
+# Remove self-loops
+self_loop_idx = src_id == dst_id
+not_self_loop_idx = src_id != dst_id
+self_loop_src_id = src_id[self_loop_idx]
+self_loop_dst_id = dst_id[self_loop_idx]
+self_loop_orig_id = g.edata['orig_id'][self_loop_idx]
+self_loop_etype = g.edata[dgl.ETYPE][self_loop_idx]
+src_id = src_id[not_self_loop_idx]
+dst_id = dst_id[not_self_loop_idx]
+orig_id = g.edata['orig_id'][not_self_loop_idx]
```
Comment (classicsong) [inline examples/pytorch/rgcn/experimental/write_mag.py:57]: use orig_eid?

File: tools/convert_partition.py [hunk 4]
```diff
@@ -42,28 +43,25 @@
...
-
-    for part_id in range(num_parts):
-        edge_file = '{}/p{:03}-{}_edges.txt'.format(input_dir, part_id, graph_name)
-        part_df = pd.read_csv(edge_file, sep=" ", usecols=remove_column_index, names=remove_column_name)
-        merge_df = pd.merge(part_df, removed_df, how='inner', on=["src_id", "dest_id"])
-        merge_df.to_csv(edge_file, mode='a', header=False, index=False, sep=" ")
-
-    print('All dropped edges were retained back into the partitioned files. Now partitioned files has all edges in them')
-
+    removed_df = csv.read_csv(removed_file, read_options=pyarrow.csv.ReadOptions(autogenerate_column_names=True),
```
Comment (classicsong) [inline tools/convert_partition.py:50]: Can we use numpy.loadtxt instead of csv.read_csv?

**Comment by Rhett-Ying** [comment]
This PR is picked and combined into another PR: https://github.com/dmlc/dgl/pull/3472.

**Pull request closed without merging**